jax/_src/cudnn/scaled_matmul_stablehlo.py

```diff
@@ -0,0 +1,730 @@
+  # Add dimensions from the LHS that are non contracting
+  for i, dim in enumerate(lhs_shape):
+    if i not in lhs_contract and i not in lhs_batch:
+      output_shape.append(dim)
+  # Add dimensions from the RHS that are non contracting
+  for i, dim in enumerate(rhs_shape):
+    if i not in rhs_contract and i not in rhs_batch:
+      output_shape.append(dim)
+  return tuple(output_shape)
+
+
+def cast_to_e8m0_with_rounding_up(x):
+  temp = x.astype(jnp.float32).view(jnp.uint32)
+  exp = temp >> 23
+  mant = temp & 0x7FFFFF
+  is_ru = jnp.logical_and(
+      jnp.logical_and((mant > 0), (exp != 0xFE)),
+      ~jnp.logical_and((exp == 0), (mant <= 0x400000))
+  )
+  exp = jnp.where(is_ru, exp + 1, exp)
+  new_x = exp.astype(jnp.uint8)
+  return new_x
+
+
+def e8m0_to_dtype(x, dtype):
+  temp = x.astype(jnp.uint32)
+  exp = temp << 23
+  new_x = exp.view(jnp.float32)
+  near_zero_value = 2**-15 if dtype == jnp.float16 else 2**-127
+  new_x = jnp.where(
+      new_x == 0, jnp.array(near_zero_value, jnp.float32), new_x
+  )
+  return new_x.astype(dtype)
+
+def quantize(x, config):
+  x_shape = x.shape
+  contract_dim = x_shape[-1]
+  block_size = config.block_size
+  assert contract_dim >= block_size and contract_dim % block_size == 0
+  x_new_shape = x_shape[:-1] + (x_shape[-1] // block_size, block_size)
+  x = x.reshape(x_new_shape)  # shape = (B, M, K / block_size, block_size)
+
+  amax = jnp.max(jnp.abs(x), axis=-1, keepdims=True)
+  MAX = jnp.finfo(config.data_type).max.astype(x.dtype)
+  scales = amax / MAX  # shape = (B, M, K / block_size, 1)
+
+  if config.mode == "mxfp8":
+    assert config.scale_type == jnp.float8_e8m0fnu
+    scales_q = cast_to_e8m0_with_rounding_up(scales)
+    scaled_x = x / e8m0_to_dtype(scales_q, scales.dtype)
+  elif config.mode == "nvfp4":
+    assert config.scale_type == jnp.float8_e4m3fn
+    # shuw(TODO): Add when XLA is ready and e2m1fn is available.
+    scales_q = scales
+    scales_x = x
+
+  clipped_x = jnp.clip(scaled_x, -MAX, MAX)
+  x_q = clipped_x.astype(config.data_type)
+
+  x_q = x_q.reshape(x_shape)  # shape = (B, M, K)
+  scales_q = jnp.reshape(scales_q, scales_q.shape[:-1]).view(
+      config.scale_type
+  )
+  return x_q, scales_q
+
+
+def quantize_to_qtype(x, q_dtype, compute_dtype, scale):
+  # Explicitly cast the max values to the compute dtype to avoid unnecessary
+  # casting to FP32 during the subsequent math operations."
+  assert q_dtype in (jnp.float8_e4m3fn, )
+  dtype_max = jnp.finfo(q_dtype).max.astype(compute_dtype)
+  scaled_x = x / jnp.broadcast_to(
+      jnp.asarray(scale, dtype=compute_dtype), x.shape
+  )
+  clipped_x = jnp.clip(scaled_x, -dtype_max, dtype_max)
+  return clipped_x.astype(q_dtype)
+
+def quantize_dequantize(x, q_dtype, scale, compute_dtype):
+  qx = quantize_to_qtype(x, q_dtype, compute_dtype, scale)
+  out = qx.astype(x.dtype) * jnp.broadcast_to(
+      jnp.asarray(scale, dtype=x.dtype), qx.shape
+  )
+  return out
+
+def generate_quantized_tensors(
```

Comment:
This function is a test utility to generate random numbers. Can you move it to tests? You can replicate them in both tests. 

Same for `quantize_dequantize` above it.